metricflow/sql/optimizer/tag_required_column_aliases.py

```diff
@@ -86,17 +87,32 @@ def _search_for_expressions(
 
     @override
     def visit_cte_node(self, node: SqlCteNode) -> None:
-        raise NotImplementedError
+        select_statement = node.select_statement
+        # Copy the tagged aliases from the CTE to the SELECT since when visiting a SELECT, the CTE node (not the SELECT
+        # in the CTE) was tagged with the required aliases.
+        required_column_aliases_in_this_node = self._column_alias_tagger.get_tagged_aliases(node)
+        self._column_alias_tagger.tag_aliases(select_statement, required_column_aliases_in_this_node)
+        # Visit parent nodes.
+        select_statement.accept(self)
 
     def _visit_parents(self, node: SqlQueryPlanNode) -> None:
         """Default recursive handler to visit the parents of the given node."""
         for parent_node in node.parent_nodes:
             parent_node.accept(self)
         return
 
+    def _tag_potential_cte_node(self, table_name: str, column_aliases: Set[str]) -> None:
+        """A reference to a SQL table might be a CTE. If so, tag the appropriate aliases in the CTEs."""
+        cte_node = self._cte_alias_to_cte_node.get(table_name)
+        if cte_node is not None:
+            self._column_alias_tagger.tag_aliases(cte_node, column_aliases)
+            # Propagate the required aliases to parents, which could be other CTEs.
+            cte_node.accept(self)
```

Comment:
Shouldn't CTEs be considered parent nodes? Why is this handled here instead of in `_visit_parents()`?

Reply:
This comment is not quite correct, so let me update. But CTE nodes are only the parents of the top level SELECT statement. So for something like:

```
-- node_id=cte_0
WITH some_table AS (
  SELECT 1
)

-- node_id=ss_0
SELECT * FROM (
  -- node_id=ss_1
  SELECT * 
  FROM 
   --node_id=st_0
    some_table
) a
```

The parents of `ss_0` are `[ss_1, cte_0]`, but the only parent of `ss_1` is `st_0`.